

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Pipeline flow &mdash; MIRZA-G 2.0.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="MIRZA-G 2.0.0 documentation" href="index.html"/>
        <link rel="prev" title="Usage" href="usage.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> MIRZA-G
          

          
          </a>

          
            
            
              <div class="version">
                2.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Instalation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Pipeline flow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prepare-mirnas-for-pipeline">1. Prepare miRNAs for pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#chunk-utrs">2. Chunk UTRs</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generate-mirna-expressions">3. Generate miRNA expressions</a></li>
<li class="toctree-l2"><a class="reference internal" href="#a-run-mirza-analysis">4a. Run MIRZA analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#i-calculate-coordinates-with-mirza">I. Calculate coordinates with MIRZA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ii-merge-and-filter-results">II. Merge and filter results</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#b-run-seed-scan">4b. Run seed scan</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-features-analysis">5. Run features analysis</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#i-calculate-mirza">I. Calculate MIRZA</a></li>
<li class="toctree-l3"><a class="reference internal" href="#ii-calculate-accessibility-with-contrafold">II. Calculate accessibility with CONTRAfold</a></li>
<li class="toctree-l3"><a class="reference internal" href="#iii-calculate-flanks-composition">III. Calculate flanks composition</a></li>
<li class="toctree-l3"><a class="reference internal" href="#iv-calculate-distance-to-the-boundary">IV. Calculate distance to the boundary</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#merge-and-add-probabilities">6. Merge and add probabilities</a></li>
<li class="toctree-l2"><a class="reference internal" href="#collect-results">7. Collect results</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">MIRZA-G</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Pipeline flow</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/scripts.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pipeline-flow">
<h1>Pipeline flow<a class="headerlink" href="#pipeline-flow" title="Permalink to this headline">¶</a></h1>
<div class="section" id="prepare-mirnas-for-pipeline">
<h2>1. Prepare miRNAs for pipeline<a class="headerlink" href="#prepare-mirnas-for-pipeline" title="Permalink to this headline">¶</a></h2>
<p>Prepare miRNA fasta file for MIRZA i.e. for each miRNA sequence
check if it is 21 nucleotide long and if not eliminate it. It
also replaces all u or U into T. In the same time it splits miRNAs
into separate files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_prepare_mirnas_for_mirza_and_split</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="o">--</span><span class="nb">input</span> <span class="n">INPUT</span>
                                             <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span> <span class="n">OUTPUT_DIR</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--input</span></kbd></td>
<td>Input miRNA file in fasta format.</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--output-dir=Output</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Directory for split files, defaults to Output</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="chunk-utrs">
<h2>2. Chunk UTRs<a class="headerlink" href="#chunk-utrs" title="Permalink to this headline">¶</a></h2>
<p>Take UTRs and generate fragments by sliding window that can be fed into MIRZA</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_generate_utr_chunks</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">input</span> <span class="n">INPUT</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="nb">dir</span>
                              <span class="n">OUTPUT_DIR</span> <span class="p">[</span><span class="o">--</span><span class="n">part</span><span class="o">-</span><span class="n">size</span> <span class="n">PART_SIZE</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">window</span><span class="o">-</span><span class="n">size</span> <span class="n">WINDOW_SIZE</span><span class="p">]</span>
                              <span class="p">[</span><span class="o">--</span><span class="n">slide</span><span class="o">-</span><span class="n">size</span> <span class="n">SLIDE_SIZE</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--input=&lt;open file '&lt;stdin&gt;', mode 'r' at 0x7f4bccea90c0&gt;</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Input file in fasta format. Defaults to sys.stdin.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--output-dir</span></kbd></td>
<td>Output directory for split files</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--part-size=40000</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Number of sequences per part, defaults to 40000</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--window-size=50</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Length of the window for MIRZA, defaults to 50</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--slide-size=20</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Size of the window slide, defaults to 20</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="generate-mirna-expressions">
<h2>3. Generate miRNA expressions<a class="headerlink" href="#generate-mirna-expressions" title="Permalink to this headline">¶</a></h2>
<p>Generate expressions of miRNAs. This file is required by MIRZA and is composed just from the
miRNA ID and its expression. Here we set up expression for each miRNA to 1.</p>
<p>There is no special script dedicated to this function. It is just the bash command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>cat input <span class="p">|</span> ruby -ne <span class="s1">&#39;puts &quot;#{$_.rstrip()[1..-1]}\t1&quot; if $_.start_with?(&quot;&gt;&quot;)&#39;</span> &gt; output
</pre></div>
</div>
</div>
<div class="section" id="a-run-mirza-analysis">
<h2>4a. Run MIRZA analysis<a class="headerlink" href="#a-run-mirza-analysis" title="Permalink to this headline">¶</a></h2>
<p>If the option &#8220;scan&#8221; as a miRNA target search is chosen putative targets are selected based on MIRZA interaction
energy. Thus, this is the  part of the pipeline scans the 3&#8217;UTRs with MIRZA to find it.</p>
<p>This script generates the jobs of the pipeline that will do the analysis on the split files.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">run_mirza_scan</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="o">--</span><span class="n">config</span> <span class="n">CONFIG</span> <span class="o">--</span><span class="n">group</span><span class="o">-</span><span class="nb">id</span> <span class="n">GROUP_ID</span>
                      <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="nb">dir</span> <span class="n">INPUT_DIR</span> <span class="o">--</span><span class="n">working</span><span class="o">-</span><span class="nb">dir</span> <span class="n">WORKING_DIR</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--config</span></kbd></td>
<td>Config file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--group-id</span></kbd></td>
<td>Group Id</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--input-dir</span></kbd></td>
<td>Input and output directory</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--working-dir</span></kbd></td>
<td>Working directory of the pipeline. Required because this file is launched from ~</td></tr>
</tbody>
</table>
</dd>
</dl>
<div class="section" id="i-calculate-coordinates-with-mirza">
<h3>I. Calculate coordinates with MIRZA<a class="headerlink" href="#i-calculate-coordinates-with-mirza" title="Permalink to this headline">¶</a></h3>
<p>Here the MIRZA algorithm is used to calculate the energy between miRNA and the 3&#8217;UTR fragments generated
before. MIRZA is launched from the command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>MIRZA expressions.tab mrnas.fa mirnas.fa <span class="m">50</span> noupdate
</pre></div>
</div>
<p>And the result is piped to the analysis script:</p>
<p>Take MIRZA output and arrange it in proper way</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_extract_data_from_mirza_output</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span><span class="p">]</span> <span class="o">--</span><span class="n">seqs</span>
                                         <span class="n">SEQS</span> <span class="o">--</span><span class="n">threshold</span> <span class="n">THRESHOLD</span>
                                         <span class="p">[</span><span class="o">--</span><span class="n">context</span> <span class="n">CONTEXT</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--output</span></kbd></td>
<td>Output file in Tab format.</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seqs</span></kbd></td>
<td>UTR sequences in fasta format</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--threshold</span></kbd></td>
<td>Threshold for the score</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--context=50</span></kbd></td>
<td>Context for sequence to print, defaults to 50</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="ii-merge-and-filter-results">
<h3>II. Merge and filter results<a class="headerlink" href="#ii-merge-and-filter-results" title="Permalink to this headline">¶</a></h3>
<p>The results are merged with bash command and duplicate results resulting from eg.
overlapping 3&#8217;UTR fragments or transcripts from the same gene.</p>
<p>Filter duplicates in coordinates by id, miRNA and sequence</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_filter_duplicates_from_scan</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="o">--</span><span class="n">coords</span> <span class="n">COORDS</span> <span class="o">--</span><span class="n">split</span><span class="o">-</span><span class="n">by</span>
                                      <span class="n">SPLIT_BY</span> <span class="o">--</span><span class="n">index</span><span class="o">-</span><span class="n">after</span><span class="o">-</span><span class="n">split</span>
                                      <span class="n">INDEX_AFTER_SPLIT</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--coords</span></kbd></td>
<td>Coordinates</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--split-by</span></kbd></td>
<td>Split id by the string</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--index-after-split</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>After split take this column as new id, 0 based</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--output</span></kbd></td>
<td>Output name</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
</div>
<div class="section" id="b-run-seed-scan">
<h2>4b. Run seed scan<a class="headerlink" href="#b-run-seed-scan" title="Permalink to this headline">¶</a></h2>
<p>This part is launched if the &#8220;seed&#8221; options is chosen when starting the pipeline. The putative
miRNA targets are found by simple seed match.</p>
<p>Count miRNA seed in the provided sequences according to chosen definition</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_count_miRNA_seeds_and_filter_duplicates</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">motifs</span> <span class="n">MOTIFS</span><span class="p">]</span>
                                                  <span class="p">[</span><span class="o">--</span><span class="n">seqs</span> <span class="n">SEQS</span><span class="p">]</span>
                                                  <span class="p">[</span><span class="o">--</span><span class="n">how</span> <span class="p">{</span><span class="n">ElMMo</span><span class="p">,</span><span class="n">TargetScan</span><span class="p">,</span><span class="mi">6</span><span class="o">-</span><span class="n">mer</span><span class="p">}]</span>
                                                  <span class="p">[</span><span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span><span class="p">]</span>
                                                  <span class="p">[</span><span class="o">--</span><span class="n">context</span> <span class="n">CONTEXT</span><span class="p">]</span>
                                                  <span class="o">--</span><span class="n">split</span><span class="o">-</span><span class="n">by</span> <span class="n">SPLIT_BY</span>
                                                  <span class="o">--</span><span class="n">index</span><span class="o">-</span><span class="n">after</span><span class="o">-</span><span class="n">split</span>
                                                  <span class="n">INDEX_AFTER_SPLIT</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--motifs</span></kbd></td>
<td>miRNA/siRNA sequences to use when scanning</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seqs</span></kbd></td>
<td>Sequences for scaning eg. 3&#8217; UTRs</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--how=TargetScan</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td><p class="first">What definition for seed to use</p>
<p class="last">Possible choices: ElMMo, TargetScan, 6-mer</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--output=coords.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Name of the output file , defaults to coords.tab</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--context=50</span></kbd></td>
<td>Context for sequence to print, defaults to 50</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--split-by</span></kbd></td>
<td>Split id by the string</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--index-after-split</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>After split take this column as new id, 0 based</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="run-features-analysis">
<h2>5. Run features analysis<a class="headerlink" href="#run-features-analysis" title="Permalink to this headline">¶</a></h2>
<p>This is the main part of the pipeline where all the features are calculated.</p>
<div class="section" id="i-calculate-mirza">
<h3>I. Calculate MIRZA<a class="headerlink" href="#i-calculate-mirza" title="Permalink to this headline">¶</a></h3>
<p>Calculate MIRZA interaction energy and MIRZA-based Branch Length Score (conservation) for the provided
coordinates of putative targets.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_calculate_MIRZA</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">onlyMIRZA</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span> <span class="p">[</span><span class="o">--</span><span class="n">seq</span> <span class="n">SEQ</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">out</span> <span class="n">OUT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">coords</span> <span class="n">COORDS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">motifs</span> <span class="n">MOTIFS</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">tree</span> <span class="n">TREE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mln</span><span class="o">-</span><span class="nb">dir</span> <span class="n">MLN_DIR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">threshold</span> <span class="n">THR</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">contextLen</span> <span class="n">CONTEXTLEN</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">mirzabin</span> <span class="n">MIRZABIN</span><span class="p">]</span>
                          <span class="p">[</span><span class="o">--</span><span class="n">reforg</span> <span class="n">REFORG</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--onlyMIRZA=no</span></kbd></td>
<td><p class="first">Calculate only MIRZA score for given coordinates</p>
<p class="last">Possible choices: yes, no</p>
</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seq=seqs.fa</span></kbd></td>
<td>Fasta with mRNA sequences</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--out=output.tab</span>, <span class="option">-o=output.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Output table</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--coords=coords.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>File with target sites positions, miRNA and target gene ID</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--motifs=</span></kbd></td>
<td>Fasta file with miRNA sequences</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--tree=</span></kbd></td>
<td>Phylogenetic tree of the species used in alignment file</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--mln-dir=</span></kbd></td>
<td>Directory with multiple alignment files</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--threshold=20.0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Threshold for MIRZA score</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--contextLen=50</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Length of the context sequence serounding binding site</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--mirzabin=</span></kbd></td>
<td>Path to the MIRZA binary</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--reforg=hg19</span></kbd></td>
<td>Reference organism to which alignments are performed: default: hg19</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="ii-calculate-accessibility-with-contrafold">
<h3>II. Calculate accessibility with CONTRAfold<a class="headerlink" href="#ii-calculate-accessibility-with-contrafold" title="Permalink to this headline">¶</a></h3>
<p>A <a class="reference external" href="http://contra.stanford.edu/contrafold/">CONTRAfold</a> algorithm is used to calculate accessibility of target site for miRNA.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_calculate_contrafold</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">seq</span> <span class="n">SEQ</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">out</span> <span class="n">OUT</span><span class="p">]</span>
                               <span class="p">[</span><span class="o">--</span><span class="n">coords</span> <span class="n">COORDS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">contextLen_L</span> <span class="n">CONTEXTLEN_L</span><span class="p">]</span>
                               <span class="p">[</span><span class="o">--</span><span class="n">contextLen_U</span> <span class="n">CONTEXTLEN_U</span><span class="p">]</span>
                               <span class="p">[</span><span class="o">--</span><span class="n">context</span> <span class="n">CONTEXT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">contrabin</span> <span class="n">CONTRABIN</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seq=seqs.fa</span></kbd></td>
<td>Fasta file with mRNA sequences , defaults to seqs.fa</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--out=output.tab.gz</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>output file, defaults to output.tab.gz</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--coords=coords.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>file with target                  sites positions, miRNA and target gene ID</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--contextLen_L=0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>length of the                   context sequence downstream binding site to be unwinded</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--contextLen_U=0</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>length of the                   context sequence upstream binding site to be unwinded</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--context=50</span></kbd></td>
<td>length of the                   context of the seed to be checked</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--contrabin=contrafold</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Path to CONTRAfold binary</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="iii-calculate-flanks-composition">
<h3>III. Calculate flanks composition<a class="headerlink" href="#iii-calculate-flanks-composition" title="Permalink to this headline">¶</a></h3>
<p>Calculate flanks composition using provided coordinates.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_calculate_flanks_composition</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">seq</span> <span class="n">SEQ</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">out</span> <span class="n">OUT</span><span class="p">]</span>
                                       <span class="p">[</span><span class="o">--</span><span class="n">coords</span> <span class="n">COORDS</span><span class="p">]</span>
                                       <span class="p">[</span><span class="o">--</span><span class="n">contextLen</span> <span class="n">CONTEXTLEN</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seq=seqs.fa</span></kbd></td>
<td>fasta with mRNA sequences</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--out=output.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>output table</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--coords=coords.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>file with target                  sites positions, miRNA and target gene ID</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--contextLen=50</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>length of the                      context sequence serounding binding site</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="iv-calculate-distance-to-the-boundary">
<h3>IV. Calculate distance to the boundary<a class="headerlink" href="#iv-calculate-distance-to-the-boundary" title="Permalink to this headline">¶</a></h3>
<p>Calculate distance to the boundary based on provided
coordinates for targets.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_calculate_distance</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">seq</span> <span class="n">SEQ</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">out</span> <span class="n">OUT</span><span class="p">]</span>
                             <span class="p">[</span><span class="o">--</span><span class="n">coords</span> <span class="n">COORDS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">contextLen</span> <span class="n">CONTEXTLEN</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--seq=seqs.fa</span></kbd></td>
<td>fasta with mRNA sequences</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--out=output.tab</span>, <span class="option">-o=output.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>output table</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--coords=coords.tab</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>file with target                  sites positions, miRNA and target gene ID</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--contextLen=50</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>length of the         context sequence serounding binding site</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
</div>
<div class="section" id="merge-and-add-probabilities">
<h2>6. Merge and add probabilities<a class="headerlink" href="#merge-and-add-probabilities" title="Permalink to this headline">¶</a></h2>
<p>Merge all results into one features table</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">rg_merge_results_add_probability_and_calculate_per_gene_score</span>
       <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">-</span><span class="n">v</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">inputs</span> <span class="n">INPUTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">coords</span> <span class="n">COORDS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="n">bls</span> <span class="n">MODEL_BLS</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">model</span><span class="o">-</span><span class="n">nobls</span> <span class="n">MODEL_NOBLS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">only</span><span class="o">-</span><span class="n">mirza</span> <span class="p">{</span><span class="n">yes</span><span class="p">,</span><span class="n">no</span><span class="p">}]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">threshold</span> <span class="n">THRESHOLD</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">split</span><span class="o">-</span><span class="n">by</span> <span class="n">SPLIT_BY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">column</span> <span class="n">COLUMN</span><span class="p">]</span>
       <span class="p">[</span><span class="o">--</span><span class="n">name</span> <span class="n">NAME</span><span class="p">]</span>
</pre></div>
</div>
<dl class="docutils">
<dt>Options:</dt>
<dd><table class="first last docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-v=False</span>, <span class="option">--verbose=False</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Be loud!</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--inputs</span></kbd></td>
<td>Coma-separated list of input paths</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--coords</span></kbd></td>
<td>Cooridinate file  used in the beginning</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--model-bls</span></kbd></td>
<td>Path to model with branch length score</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--model-nobls</span></kbd></td>
<td>Path to model without branch length score</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--output</span></kbd></td>
<td>Output file name</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--only-mirza</span></kbd></td>
<td><p class="first">Calculate only MIRZA and DON&#8217;T calculate MIRZA BLS</p>
<p class="last">Possible choices: yes, no</p>
</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--threshold=0.12</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Threshold for summing, defaults to 0.12</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">--split-by=NONE</span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>If the header of fasta has multiple annotations eg. transcript_id|entrez_id|wikiname
split it and take only one, defaults to NONE</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--column=0</span></kbd></td>
<td>0 based column number to take after splitting, defaults to 0</td></tr>
<tr><td class="option-group">
<kbd><span class="option">--name=GeneID</span></kbd></td>
<td>Name of the id eg. gene, transcript etc , defaults to GeneID</td></tr>
</tbody>
</table>
</dd>
</dl>
</div>
<div class="section" id="collect-results">
<h2>7. Collect results<a class="headerlink" href="#collect-results" title="Permalink to this headline">¶</a></h2>
<p>In the end all the results are collected with bash command:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>zcat output_dir/<span class="se">\*</span>.score &gt; cwd/mirza_g_results_protocol.tab
</pre></div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href="usage.html" class="btn btn-neutral" title="Usage" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Rafal Gumienny.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'2.0.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>